webapp/src/Controller/Jury/SubmissionController.php

```diff
@@ -900,30 +886,70 @@ public function sourceAction(
                 ->getQuery()
                 ->getOneOrNullResult();
         }
+        if ($oldSubmission !== null) {
+            $otherSubmissions[] = [
+                'submitid' => $oldSubmission->getSubmitid(),
+                'tag'      => 'previous',
+            ];
+        }
 
-        /** @var SubmissionFile[] $files */
+        $files_query = array_map(fn($s) => $s['submitid'], $otherSubmissions);
+        $files_query[] = $submission->getSubmitid();
+        /** @var SubmissionFile[] $oldFiles */
         $oldFiles = $this->em->createQueryBuilder()
             ->from(SubmissionFile::class, 'file')
             ->select('file')
-            ->andWhere('file.submission = :submission')
-            ->setParameter('submission', $oldSubmission)
-            ->orderBy('file.ranknumber')
+            ->andWhere('file.submission in (:submissions)')
+            ->setParameter('submissions', $files_query)
+            ->orderBy('file.submission, file.ranknumber')
             ->getQuery()
             ->getResult();
 
-        $oldFileStats      = $oldFiles !== null ? $this->determineFileChanged($files, $oldFiles) : [];
-        $originalFileStats = $originalFiles !== null ? $this->determineFileChanged($files, $originalFiles) : [];
+        /** @var array<string, array<int, array{
+         *      rank: int,
+         *      filename: string,
+         *      source: string,
+         *      renamedFrom?: string
+         * }>> $files */
+        $files = [];
+        /** @var (string|false)[] $renames */
+        $renames = [];
+        foreach ($oldFiles as $f) {
+            $submitId = $f->getSubmission()->getSubmitid();
+            $files[$f->getFilename()] ??= [];
+            $files[$f->getFilename()][$submitId] = [
+                'rank' => $f->getRank(),
+                'filename' => $f->getFilename(),
+                'source'   => mb_check_encoding($f->getSourcecode(), 'UTF-8') ? $f->getSourcecode() : "Could not display binary file",
+            ];
+
+            // Keep track of the single filename within a submission for handling renaming.
+            $renames[$submitId] = array_key_exists($submitId, $renames) ? false : $f->getFilename();
```

Comment:
So here it would almost make sense to set the $renames as:
```@var array<int, (string|false)>```

because the id is important,